lib/Conversion/SeqToSV/FirRegLowering.cpp

```diff
@@ -19,6 +20,33 @@ using llvm::MapVector;
 
 #define DEBUG_TYPE "lower-seq-firreg"
 
+// Reimplemented from SliceAnalysis to use a worklist rather than recursion and
+// non-insert ordered set.
+static void getForwardSliceSimple(Operation *root,
+                                  SmallPtrSetImpl<Operation *> &forwardSlice) {
+  SmallVector<Operation *> worklist({root});
+
+  while (!worklist.empty()) {
+    Operation *op = worklist.pop_back_val();
+
+    if (!op)
+      continue;
+
+    for (Region &region : op->getRegions())
+      for (Block &block : region)
+        for (Operation &blockOp : block)
+          if (!forwardSlice.contains(&blockOp))
```

Comment:
Guard insertion to worklist by `if (!forwardSlice.insert(&blockOp).second)` -- to save a lot of possibly redundant entries? Especially for the user walk below, I'd expect.

Otherwise you can add the same operation MANY times before visiting it in the worklist (and then walking it a bunch it looks like)?

Reply:
Thanks for the tip, just pushed a commit with that suggestion and testing it out on a real design.